internal/services/controller/controller_test.go

```diff
@@ -242,6 +243,157 @@ func TestController_ApiResourcesErrorProcessing(t *testing.T) {
 	}])
 }
 
+func TestController_ShouldSendByInterval(t *testing.T) {
+	tt := []struct {
+		name          string
+		sendInterval  time.Duration
+		sendDurations []time.Duration
+		checkAfter    time.Duration
+		wantSends     int64
+	}{
+		{
+			name:         "should trigger all sends when none exceed allocated intervals",
+			sendInterval: 300 * time.Millisecond,
+			sendDurations: []time.Duration{
+				// Send by 300ms
+				250 * time.Millisecond,
+				// ...by 600ms
+				250 * time.Millisecond,
+				// ...by 900ms
+				250 * time.Millisecond,
+				// ...by 1200ms
+				250 * time.Millisecond,
+			},
+			checkAfter: 1200 * time.Millisecond,
+			wantSends:  4,
+		},
+		{
+			name:         "should trigger all sends when previous send exceeds one interval",
+			sendInterval: 300 * time.Millisecond,
+			sendDurations: []time.Duration{
+				// At 0, 300: idle, previous still sending
+				// Send by 600 ms
+				450 * time.Millisecond,
+				// Send by 600 ms too, since we expect ticker to fire events even between ticks
+				50 * time.Millisecond,
+			},
+			checkAfter: 600 * time.Millisecond,
+			wantSends:  2,
+		},
+		{
+			name:         "should trigger all sends when previous send exceeds two intervals",
+			sendInterval: 300 * time.Millisecond,
+			sendDurations: []time.Duration{
+				// At 0, 300, 600ms: idle, previous still sending
+				// Send by 900 ms
+				650 * time.Millisecond,
+				// Send by 900 ms too, since we expect ticker to fire events even between ticks
+				150 * time.Millisecond,
+			},
+			checkAfter: 900 * time.Millisecond,
+			wantSends:  2,
+		},
+	}
+
+	for _, tc := range tt {
+		t.Run(tc.name, func(t *testing.T) {
+			mockctrl := gomock.NewController(t)
+			castaiclient := mock_castai.NewMockClient(mockctrl)
+			version := mock_version.NewMockInterface(mockctrl)
+			provider := mock_types.NewMockProvider(mockctrl)
+
+			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancel()
+
+			pod := &v1.Pod{ObjectMeta: metav1.ObjectMeta{Namespace: v1.NamespaceDefault, Name: "pod1"}}
+			_, err := delta.Encode(pod)
+			require.NoError(t, err)
+
+			clientset := fake.NewSimpleClientset()
+			metricsClient := metrics_fake.NewSimpleClientset()
+			dynamicClient := dynamic_fake.NewSimpleDynamicClient(runtime.NewScheme())
+
+			version.EXPECT().Full().Return("1.21+").AnyTimes()
+
+			clusterID := uuid.New()
+			log := logrus.New()
+
+			var gotSends atomic.Int64
+			var wg sync.WaitGroup
+			wg.Add(len(tc.sendDurations))
+			var sentFirstTickNotification atomic.Bool
+
+			var firstTickAt time.Time
+			var lastSentAt time.Time
+
+			for _, sendDuration := range tc.sendDurations {
+				castaiclient.EXPECT().
+					SendDelta(gomock.Any(), clusterID.String(), gomock.Any()).
+					DoAndReturn(func(_ context.Context, clusterID string, d *castai.Delta) error {
+						if !sentFirstTickNotification.Load() {
```

Comment:
Using atomic bool here suggests that this func can be called concurrently at the same time, but there's a race condition. If two deltas are sent at the same time, they can both get `false` from `Load()` and will then race to set the `firstTickAt`